types/+MPCT/code_MPCT_ADMM_semiband_C.c

```diff
@@ -0,0 +1,930 @@
+
+        res_flag = 0; // Reset the residual flag
+
+        for (unsigned int i = 0 ; i < (NN_+1)*nm_ ; i++){
+            
+            res_fixed_point = v[i] - v_old[i];
+            res_primal_feas = z[i] - v[i];
+            // Obtain absolute values
+            res_fixed_point = (res_fixed_point > 0.0) ? res_fixed_point : -res_fixed_point;
+            res_primal_feas = (res_primal_feas > 0.0) ? res_primal_feas : -res_primal_feas;
+            
+            if (res_fixed_point > tol || res_primal_feas > tol){
+
+                res_flag = 1;
+                break;
+
+            }
+
+        }
+
+        // Exit condition
+        if (res_flag == 0){
+
+            done = 1;
+            *e_flag = 1;
+
+        }
+        else if (k >= k_max){
+
+            done = 1;
+            *e_flag = -1;        
+
+        }
+
+    }
+
+    // Measure time
+    #if MEASURE_TIME == 1
+    
+    #if WIN32
+    QueryPerformanceCounter(&post_solve); // Get time after solving
+    t_solve_time = 1000000000ULL * (post_solve.QuadPart - post_update.QuadPart) / frequency.QuadPart;
+    sol->solve_time = t_solve_time/(double)1e+9;
+    #else // If Linux
+    clock_gettime(CLOCK_MONOTONIC_RAW, &post_solve);
+    sol->solve_time = (double) ( (post_solve.tv_sec - post_update.tv_sec) * 1000.0 ) + (double) ( (post_solve.tv_nsec - post_update.tv_nsec) / 1000000.0 );
+    #endif
+    
+    #endif
+
+    // Control action
+    #if in_engineering == 1
+    for (unsigned int i = nn_ ; i < nm_ ; i++){
+
+        u_opt[i-nn_] = v[i] * scaling_i_u[i-nn_] + OpPoint_u[i-nn_];
+
+    }
+    #endif
+    #if in_engineering == 0
+    for (unsigned int i = nn_ ; i < nm_ ; i++){
+
+        u_opt[i-nn_] = v[i];
+
+    }
+    #endif
+
+    // Return number of iterations
+    *k_in = k;
+
+    // Save solution into structure
+    #ifdef DEBUG
+
+    for (unsigned int i = 0 ; i < (NN_+1)*nm_ ; i++){
+        
+        sol->z[i] = z[i];
+        sol->v[i] = v[i];
+        sol->lambda[i] = lambda[i];
+    
+    }
+
+    #endif
+
+    // Measure time
+    #if MEASURE_TIME == 1
+    
+    #if WIN32
+    QueryPerformanceCounter(&post_polish); // Get time after polishing
+    t_run_time = 1000000000ULL * (post_polish.QuadPart - start.QuadPart) / frequency.QuadPart;
+    t_polish_time = 1000000000ULL * (post_polish.QuadPart - post_solve.QuadPart) / frequency.QuadPart;
+    sol->run_time = t_run_time/(double)1e+9;
+    sol->polish_time = t_polish_time/(double)1e+9;
+    #else // If Linux
+    clock_gettime(CLOCK_MONOTONIC_RAW, &post_polish);
+    sol->run_time = (double) ( (post_polish.tv_sec - start.tv_sec) * 1000.0 ) + (double) ( (post_polish.tv_nsec - start.tv_nsec) / 1000000.0 );
+    sol->polish_time = (double) ( (post_polish.tv_sec - post_solve.tv_sec) * 1000.0 ) + (double) ( (post_polish.tv_nsec - post_solve.tv_nsec) / 1000000.0 );
+    #endif
+    
+    #endif
+
+}
+
+
+
+void solve_banded_Chol(const double (*Alpha)[nn_][nn_], const double (*Beta)[nn_][nn_], double *z, double *d){
```

Comment:
This function was already programmed. It is essentially function `solve_W_matrix_form()` from `code_laxMPC_FISTA_C.c`. The existing implementation does not need a `sum` variable.  Could it be avoided here?

This function is the most computationally demanding of the algorithm. So we need to make sure that it's implementation is as efficient as it can be. For instance, we should avoid the `if (k > 0)`.

Reply:
Fixed